subprojects/GOPATH/src/github.com/microsoft/hivedscheduler/pkg/algorithm/topology_aware_scheduler.go

```diff
@@ -0,0 +1,284 @@
+		// free GPUs will be used first (before preemptible GPUs)
+		availableGpus = append(availableGpus, preemptibleGpus...)
+	}
+	availableGpuIndex := int32(0)
+	searchGpuIndex := int32(0)
+	var gpu Cell
+	for {
+		for availableGpuIndex < int32(len(availableGpus)) {
+			gpu = availableGpus[availableGpuIndex]
+			t.currentSolutionIndices[searchGpuIndex] = availableGpuIndex
+			if searchGpuIndex == 0 {
+				t.currentLCA[searchGpuIndex] = gpu
+			} else {
+				t.currentLCA[searchGpuIndex] = findLCA(gpu, t.currentLCA[searchGpuIndex-1])
+				// pruning: if the current LCA has been higher than the lowest ever,
+				// the node will be skipped
+				if (t.currentLCA[searchGpuIndex] == nil && t.lowestLCALevel < highestLevel) ||
+					(t.currentLCA[searchGpuIndex].GetLevel() > t.lowestLCALevel) {
+					availableGpuIndex++
+					continue
+				}
+			}
+			if searchGpuIndex == gpuNumber-1 {
+				if t.checkCurrentGpus(availableGpus) {
+					// early stop: return if the solution is optimal (i.e., all buddies)
+					availableGpus = t.removePickedGpus(availableGpus)
+					return availableGpus
+				}
+			} else {
+				searchGpuIndex++
+			}
+			availableGpuIndex++
+		}
+		searchGpuIndex--
+		if searchGpuIndex < 0 {
+			if t.lowestLCALevel == highestLevel {
+				panic(fmt.Sprintf("failed to allocate %v GPUs in picked node %v", gpuNumber, node.GetName()))
+			}
+			availableGpus = t.removePickedGpus(availableGpus)
+			return availableGpus
+		}
+		availableGpuIndex = t.currentSolutionIndices[searchGpuIndex] + 1
+	}
+}
+
+// checkCurrentGpus checks if the currently picked GPUs have the lowest LCA. It also checks if the solution
+// is optimal (if the GPUs are all buddies).
+func (t *topologyAwareScheduler) checkCurrentGpus(gpus CellList) bool {
+	currentLCALevel := t.currentLCA[len(t.currentLCA)-1].GetLevel()
+	if currentLCALevel < t.lowestLCALevel {
+		copy(t.lowestLCASolutionIndices, t.currentSolutionIndices)
+		for i := 0; i < len(t.currentSolutionIndices); i++ {
+			t.lowestLCASolution[i] = gpus[t.currentSolutionIndices[i]]
+		}
+		t.lowestLCALevel = currentLCALevel
+		if t.lowestLCALevel == t.lowestLCASolution[0].GetLevel()+1 {
+			return true
+		}
+	}
+	return false
+}
+
+// removePickedGpus remove picked GPUs from the available GPU list.
+func (t *topologyAwareScheduler) removePickedGpus(gpus CellList) CellList {
+	for i, index := range t.lowestLCASolutionIndices {
+		offset := int32(i)
+		if i < len(t.lowestLCASolutionIndices)-1 {
+			nextIndex := t.lowestLCASolutionIndices[i+1]
+			copy(gpus[index-offset:nextIndex-offset-1], gpus[index+1:nextIndex])
+		} else {
+			copy(gpus[index-offset:], gpus[index+1:])
+		}
+	}
+	return gpus[:len(gpus)-len(t.currentSolutionIndices)]
+}
+
+func sortGpuNumbers(n []int32) {
+	tmp := make([]int, len(n))
+	for i := 0; i < len(n); i++ {
+		tmp[i] = int(n[i])
+	}
+	sort.Ints(tmp)
+	for i := 0; i < len(tmp); i++ {
+		n[i] = int32(tmp[i])
+	}
+}
```

Comment:
Why not directly use sort.Ints(tmp)

Reply:
sort.Ints([]int) only accepts []int, not []int32... Go does not have build-in sort for int32. :(

---
In reply to: [332855059](https://github.com/microsoft/pai/pull/3684#discussion_r332855059) [](ancestors = 332855059)